src/gc.c

```diff
@@ -108,7 +108,7 @@ static void run_finalizer(jl_ptls_t ptls, jl_value_t *o, jl_value_t *ff)
     JL_TRY {
         size_t last_age = jl_get_ptls_states()->world_age;
         jl_get_ptls_states()->world_age = jl_world_counter;
-        jl_apply(args, 2);
+        jl_apply_with_saved_exception_state(args, 2, 0);
         jl_get_ptls_states()->world_age = last_age;
     }
     JL_CATCH {
```

Comment:
> in case a finalizer catches and hides an exception.

Ah, yeah.

> C is just not good for this kind of programming :(

Agreed. Luckily, there's only one caller so it should be sufficient to add the saving logic to `jl_gc_run_finalizers_in_lilst` after releasing the finalizer lock.